stream_alert/classifier/clients/firehose.py

```diff
@@ -0,0 +1,410 @@
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from collections import defaultdict
+import io
+import json
+import re
+
+import backoff
+import boto3
+from botocore.exceptions import ClientError
+from botocore.vendored.requests.exceptions import ConnectionError, Timeout
+import jsonlines
+
+from stream_alert.shared import CLASSIFIER_FUNCTION_NAME as FUNCTION_NAME
+from stream_alert.shared.helpers import boto
+from stream_alert.shared.logger import get_logger
+from stream_alert.shared.metrics import MetricLogger
+from stream_alert.shared.backoff_handlers import (
+    backoff_handler,
+    giveup_handler,
+    success_handler
+)
+
+LOGGER = get_logger(__name__)
+
+
+class FirehoseClient(object):
+    """Handles preparing and sending data from the Rule Processor to Kinesis Firehose"""
+    # Used to detect special characters in payload keys.
+    # This is necessary for sanitization of data prior to searching in Athena.
+    SPECIAL_CHAR_REGEX = re.compile(r'\W')
+    SPECIAL_CHAR_SUB = '_'
+    # For PutRecordBatch backoff
+    MAX_BACKOFF_ATTEMPTS = 10
+    # Adds a max of 20 seconds more to the Lambda function
+    MAX_BACKOFF_FIBO_VALUE = 8
+    # Set Firehose Limits: http://bit.ly/2fw5UY2
+    MAX_BATCH_COUNT = 500
+    MAX_BATCH_SIZE = 4000 * 1000
+    # The subtraction of 2 accounts for the newline at the end
+    MAX_RECORD_SIZE = 1000 * 1000 - 2
+
+    # Default firehose name prefix
+    DEFAULT_FIREHOSE_PREFIX = 'streamalert_data_{}'
+
+    # Exception for which backoff operations should be performed
+    EXCEPTIONS_TO_BACKOFF = (ClientError, ConnectionError, Timeout)
+
+    # Set of enabled log types for firehose, loaded from configs
+    _ENABLED_LOGS = dict()
+
+    def __init__(self, firehose_config=None, log_sources=None):
+        self._client = boto3.client('firehose', config=boto.default_config())
+        # Create a dictionary to hold parsed payloads by log type.
+        # Firehose needs this information to send to its corresponding
+        # delivery stream.
+        self._categorized_records = defaultdict(list)
+
+        self.load_enabled_log_sources(firehose_config, log_sources, force_load=True)
+
+    @classmethod
+    def _records_json_lines(cls, records):
+        """Write the dictionary records to json lines and return the list of lines
+
+        Args:
+            records (list<dict>): Records to be written to line-delimited json
+
+        Returns:
+            list: JSON serialized records
+        """
+        # Write the json lines to the object in minimal form
+        json_lines = io.StringIO()
+        with jsonlines.Writer(json_lines, compact=True) as writer:
```

Comment:
Talked offline. `jsonlines` won't bring benefit. We will switch back to `json.dumps`